In [1]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc read-aloud"><h2>--- Day 2: Inventory Management System ---</h2><p>You stop falling through time, catch your breath, and check the screen on the device. "Destination reached. Current Year: 1518. Current Location: North Pole Utility Closet 83N10." You made it! Now, to find those anomalies.</p>
<p>Outside the utility closet, you hear footsteps and a voice. "...I'm not sure either. But now that <span title="This is, in fact, roughly when chimneys became common in houses.">so many people have chimneys</span>, maybe he could sneak in that way?" Another voice responds, "Actually, we've been working on a new kind of <em>suit</em> that would let him fit through tight spaces like that. But, I heard that a few days ago, they lost the prototype fabric, the design plans, everything! Nobody on the team can even seem to remember important details of the project!"</p>
<p>"Wouldn't they have had enough fabric to fill several boxes in the warehouse? They'd be stored together, so the box IDs should be similar. Too bad it would take forever to search the warehouse for <em>two similar box IDs</em>..." They walk too far away to hear any more.</p>
<p>Late at night, you sneak to the warehouse - who knows what kinds of paradoxes you could cause if you were discovered - and use your fancy wrist device to quickly scan every box and produce a list of the likely candidates (your puzzle input).</p>
<p>To make sure you didn't miss any, you scan the likely candidate boxes again, counting the number that have an ID containing <em>exactly two of any letter</em> and then separately counting those with <em>exactly three of any letter</em>. You can multiply those two counts together to get a rudimentary <a href="https://en.wikipedia.org/wiki/Checksum">checksum</a> and compare it to what your device predicts.</p>
<p>For example, if you see the following box IDs:</p>
<ul>
<li><code>abcdef</code> contains no letters that appear exactly two or three times.</li>
<li><code>bababc</code> contains two <code>a</code> and three <code>b</code>, so it counts for both.</li>
<li><code>abbcde</code> contains two <code>b</code>, but no letter appears exactly three times.</li>
<li><code>abcccd</code> contains three <code>c</code>, but no letter appears exactly two times.</li>
<li><code>aabcdd</code> contains two <code>a</code> and two <code>d</code>, but it only counts once.</li>
<li><code>abcdee</code> contains two <code>e</code>.</li>
<li><code>ababab</code> contains three <code>a</code> and three <code>b</code>, but it only counts once.</li>
</ul>
<p>Of these box IDs, four of them contain a letter which appears exactly twice, and three of them contain a letter which appears exactly three times. Multiplying these together produces a checksum of <code>4 * 3 = 12</code>.</p>
<p><em>What is the checksum</em> for your list of box IDs?</p>
</article>


In [2]:
from collections import Counter


example = """
abcdef
bababc
abbcde
abcccd
aabcdd
abcdee
ababab
"""


def checksum(boxes: str) -> int:
    twos, threes = 0, 0
    for l in boxes.strip().splitlines():
        counts = set(Counter(l).values())
        twos += 1 if 2 in counts else 0
        threes += 1 if 3 in counts else 0

    return twos * threes


print(f"Example: {checksum(example)} should be 12")

Example: 12 should be 12


In [3]:
with open("../input/day2.txt") as f:
    puzzle = f.read()

print(f"Part I: {checksum(puzzle)}")

Part I: 3952


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>3952</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>Confident that your list of box IDs is complete, you're ready to find the boxes full of prototype fabric.</p>
<p>The boxes will have IDs which differ by exactly one character at the same position in both strings. For example, given the following box IDs:</p>
<pre><code>abcde
fghij
klmno
pqrst
fguij
axcye
wvxyz
</code></pre>
<p>The IDs <code>abcde</code> and <code>axcye</code> are close, but they differ by two characters (the second and fourth). However, the IDs <code>fghij</code> and <code>fguij</code> differ by exactly one character, the third (<code>h</code> and <code>u</code>). Those must be the correct boxes.</p>
<p><em>What letters are common between the two correct box IDs?</em> (In the example above, this is found by removing the differing character from either ID, producing <code>fgij</code>.)</p>
</article>

</main>


In [4]:
example = """
abcde
fghij
klmno
pqrst
fguij
axcye
wvxyz
"""


def get_correct_id(boxes: str) -> str:
    ids = boxes.strip().splitlines()
    wanted_length = len(ids[0]) - 1
    for id1, id2 in product(ids, repeat=2):
        in_both = [id1[i] for i in range(len(id1)) if id1[i] == id2[i]]
        if len(in_both) == wanted_length:
            return "".join(in_both)


print(f"Example: {get_correct_id(example)} should be fgij")

Example: fgij should be fgij


In [5]:
print(f"Part II: {get_correct_id(puzzle)}")

Part II: vtnikorkulbfejvyznqgdxpaw


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>vtnikorkulbfejvyznqgdxpaw</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>
